adagrams/game.py

```diff
@@ -1,11 +1,84 @@
+
+import random
 def draw_letters():
-    pass
+    distribution_of_letters = {"A" : 9, "N" : 6, "B" : 2, "O" : 8, "C" : 2, "P" : 2, "D" : 4 ,"Q" : 1, "E" : 12, "R" : 6, "F" : 2, "S" : 4, "G" : 3, "T" : 6, "H" : 2, "U" : 4, "I" : 9, "V" : 2, "J" : 1 , "W" : 2, "K" : 1, "X" : 1, "L" : 4, "Y" : 2, "M" : 2, "Z" : 1}
+    letters = []
+    while len(letters) < 10:
+        userletters = random.choice(list(distribution_of_letters.items()))
+        if userletters[1] > 0:
+            letters.append(userletters[0])
+            distribution_of_letters[userletters[0]]-=1
+    return letters
 
 def uses_available_letters(word, letter_bank):
-    pass
+    counter = {}
+    for letter in letter_bank:
+        if letter in counter:
+            counter[letter] += 1
+        else:
+            counter[letter] = 1
+    
+    for ch in word.upper():
+            if ch in counter and counter[ch] > 0:
+                counter[ch] -= 1
+            else:
+                return False
+    return True
 
 def score_word(word):
-    pass
+    point_1_letters = ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"]
+    point_2_letters = ["D", "G"]
+    point_3_letters = ["B", "C", "M", "P"]
+    point_4_letters = ["F", "H", "V", "W", "Y"]
+    point_5_letters = ["K"]
+    point_8_letters = ["J", "X"]
+    point_10_letters = ["Q", "Z"]
+    
+    # use map to store letter to score relationships
+    score_chart = {
+        1: point_1_letters,
+        2: point_2_letters,
+        3: point_3_letters,
+        4: point_4_letters,
+        5: point_5_letters,
+        8: point_8_letters,
+        10: point_10_letters
+        }
+    
+    total_score = 0
+
+    if len(word) != 0:
+        special_length = [7, 8, 9, 10]
+
+        # add additional 8 points for words with special length
+        if len(word) in special_length:
+            total_score += 8
+        
+        # find the score for each letter in word, and add score to total_score
+        for point, letters in score_chart.items():
+            for letter in word.upper():
+                if letter in letters:
+                    total_score += point
```

Comment:
this works just fine! One thing to consider: looking up an item in a list is generally O(n), but the lookup time for a dictionary is O(1).

If we kept our dictionary like: `letter_dict = {"A": 1, "B": 2....}`, then our look up will be constant, and that will get rid of one of the the three for loops in this code block.

```python
for letter in word.upper():
    if letter in letter_dict:
        # increase your total_score here
```

now our code only has one for loop in it!